candid/src/parser/value.rs

```diff
@@ -319,6 +319,78 @@ impl IDLValue {
     }
 }
 
+pub mod pretty {
+    use super::*;
+    use crate::pretty::*;
+
+    use ::pretty::RcDoc;
+
+    pub fn pp_label(id: &Label) -> RcDoc {
+        match id {
+            Label::Named(id) => str("\"")
+                .append(format!("{}", id.escape_debug()))
+                .append("\""),
+            Label::Id(n) | Label::Unnamed(n) => RcDoc::as_string(n),
+        }
+    }
+
+    pub fn pp_idl_field(field: &IDLField) -> RcDoc {
+        pp_label(&field.id)
+            .append(kwd(" ="))
+            .append(pp_value(&field.val))
+    }
+
+    pub fn pp_idl_fields(fields: &[IDLField]) -> RcDoc {
+        concat(fields.iter().map(|f| pp_idl_field(f)), ";")
+    }
+
+    pub fn pp_variant_field(field: &IDLField) -> RcDoc {
+        pp_label(&field.id)
+            .append(kwd(" ="))
+            .append(pp_value(&field.val))
+    }
+
+    pub fn pp_value(v: &IDLValue) -> RcDoc {
+        use super::IDLValue::*;
+        match &*v {
+            Bool(true) => str("true"),
+            Bool(false) => str("false"),
+            Null => str("null"),
+            None => str("none"),
+            Int(i) => RcDoc::as_string(i),
+            Nat(n) => RcDoc::as_string(n),
+            Nat8(n) => RcDoc::as_string(n),
+            Nat16(n) => RcDoc::as_string(n),
+            Nat32(n) => RcDoc::as_string(n),
+            Nat64(n) => RcDoc::as_string(n),
+            Int8(i) => RcDoc::as_string(i),
+            Int16(i) => RcDoc::as_string(i),
+            Int32(i) => RcDoc::as_string(i),
+            Int64(i) => RcDoc::as_string(i),
+            Number(t) => str(t),
+            Text(t) => RcDoc::as_string(format!("{:?}", t)), // to do -- enough quoting here?
```

Comment:
Right, this is tricky to know if we want to escape characters or not. I think it depends on application, so best to have a flag to control this.

Reply:
```suggestion
            Text(t) => RcDoc::as_string(format!("\"{:?}\"", t)), // to do -- enough quoting here?
```

Reply:
Your suggestion will give double quotes, which I don't think you want.

Notice that I'm using the Debug formatting string `{:?}`, not `{}`.

This will always include enclosing quotes.  However, it does not escape quotes in the string itself, so the string `\"` will be pretty printed as `"""`, and not `"\""`, which is what I'd want and expect.

Reply:
In any case, it seems like a separate PR to add tests that stress the parsing/pretty-printing round trips, and ensure that they always agree.

For this PR, having the pretty printer work at all is enough for an MVP with some reasonable human readable output.  No need for strong mathematical properties (yet).

Reply:
Ah, right. I forget debug print has a quote already. For `\n`, I think debug print won't make a new line. We can tune this later.